ledger/sandbox/src/main/scala/com/digitalasset/platform/apiserver/TimeServiceBackend.scala

```diff
@@ -38,19 +45,48 @@ object TimeServiceBackend {
     }
   }
 
-  private class ObservingTimeServiceBackend(
-      timeProvider: TimeServiceBackend,
-      onTimeChange: Instant => Future[Unit]
-  ) extends TimeServiceBackend {
-    override def getCurrentTime: Instant = timeProvider.getCurrentTime
-
-    override def setCurrentTime(expectedTime: Instant, newTime: Instant): Future[Boolean] =
-      timeProvider
-        .setCurrentTime(expectedTime, newTime)
-        .flatMap { success =>
-          if (success)
-            onTimeChange(expectedTime).map(_ => true)(DirectExecutionContext)
-          else Future.successful(false)
-        }(DirectExecutionContext)
+  private final class AkkaQueueBasedObservedTimeServiceBackend(delegate: TimeServiceBackend)(
+      implicit materializer: Materializer
+  ) extends ObservedTimeServiceBackend {
+    // There is no `TrieSet` type, so we emulate it with a `TrieMap` with meaningless values.
+    // Scala doesn't react well to `()` in tuples, so we're using Boolean instead of Unit here.
+    private val queues = concurrent.TrieMap[SourceQueueWithComplete[Instant], Boolean]()
+
+    override def getCurrentTime: Instant =
+      delegate.getCurrentTime
+
+    override def setCurrentTime(currentTime: Instant, newTime: Instant): Future[Boolean] =
+      delegate
+        .setCurrentTime(currentTime, newTime)
+        .andThen {
+          case Success(true) =>
+            queues.keys.foreach(_.offer(newTime))
+        }(materializer.executionContext)
+
+    override def changes: ResourceOwner[Source[Instant, NotUsed]] =
+      new ResourceOwner[Source[Instant, NotUsed]] {
+        override def acquire()(
+            implicit executionContext: ExecutionContext
+        ): Resource[Source[Instant, NotUsed]] = {
+          Resource(Future {
```

Comment:
The indentation is hard to follow here. Can you possibly break this apart into smaller bits to improve readability, please?